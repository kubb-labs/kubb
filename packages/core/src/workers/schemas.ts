import { z } from 'zod'

import type { ZodStringDef, ZodType } from 'zod'
import type { KubbFile } from '../FileManager.ts'

export const importSchema = z.object({
  name: z.union([z.string(), z.array(z.string())]),
  path: z.string(),
  isTypeOnly: z.boolean().optional(),
  isNameSpace: z.boolean().optional(),
  root: z.string().optional(),
})

export const exportSchema = z.object({
  name: z.union([z.string(), z.array(z.string())]).optional(),
  path: z.string(),
  isTypeOnly: z.boolean().optional(),
  asAlias: z.boolean().optional(),
})

export const fileSchema = z.object({
  id: z.string().optional(),
  baseName: z.string() as ZodType<KubbFile.BaseName, ZodStringDef>,
  path: z.string(),
  source: z.string(),
  imports: z.array(importSchema).optional(),
  exports: z.array(exportSchema).optional(),
  override: z.boolean().optional(),
  meta: z.object({}).optional(),
  env: z.object({}).optional(),
})
