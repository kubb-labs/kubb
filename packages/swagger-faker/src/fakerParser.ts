export type FakerMetaMapper = {
  object: { keyword: 'object'; args: { entries: { [x: string]: FakerMeta[] }; strict?: boolean } }
  url: { keyword: 'url' }
  uuid: { keyword: 'uuid' }
  email: { keyword: 'email' }
  firstName: { keyword: 'firstName' }
  lastName: { keyword: 'lastName' }
  phone: { keyword: 'phone' }
  password: { keyword: 'password' }
  datetime: { keyword: 'datetime' }
  tuple: { keyword: 'tuple'; args?: FakerMeta[] }
  array: { keyword: 'array'; args?: FakerMeta[] }
  enum: { keyword: 'enum'; args?: Array<string | number> }
  and: { keyword: 'and'; args?: FakerMeta[] }
  union: { keyword: 'union'; args?: FakerMeta[] }
  ref: { keyword: 'ref'; args?: { name: string } }
  catchall: { keyword: 'catchall'; args?: FakerMeta[] }
  matches: { keyword: 'matches'; args?: string }
  boolean: { keyword: 'boolean' }
  string: { keyword: 'string'; args?: { min?: number; max?: number } }
  integer: { keyword: 'integer'; args?: { min?: number; max?: number } }
  number: { keyword: 'number'; args?: { min?: number; max?: number } }
  undefined: { keyword: 'undefined' }
  null: { keyword: 'null' }
  any: { keyword: 'any' }
  unknown: { keyword: 'unknown' }
}

export const fakerKeywords = {
  any: 'any',
  unknown: 'unknown',
  number: 'number',
  integer: 'integer',
  string: 'string',
  boolean: 'boolean',
  undefined: 'undefined',
  null: 'null',
  array: 'array',
  tuple: 'tuple',
  enum: 'enum',
  union: 'union',
  datetime: 'datetime',
  email: 'email',
  uuid: 'uuid',
  url: 'url',
  /* intersection */
  and: 'and',

  // custom ones
  object: 'object',
  ref: 'ref',
  catchall: 'catchall',
  matches: 'matches',
  firstName: 'firstName',
  lastName: 'lastName',
  password: 'password',
  phone: 'phone',
} satisfies { [K in keyof FakerMetaMapper]: FakerMetaMapper[K]['keyword'] }

export type FakerKeyword = keyof typeof fakerKeywords

export const fakerKeywordMapper = {
  any: 'undefined',
  unknown: 'unknown',
  number: 'faker.number.float',
  integer: 'faker.number.int',
  string: 'faker.string.alpha',
  boolean: 'faker.datatype.boolean',
  undefined: 'undefined',
  null: 'null',
  array: 'faker.helpers.arrayElements',
  tuple: 'faker.helpers.arrayElements',
  enum: 'faker.helpers.arrayElement<any>',
  union: 'faker.helpers.arrayElement',
  datetime: 'faker.date.anytime',
  uuid: 'faker.string.uuid',
  url: 'faker.internet.url',
  /* intersection */
  and: 'Object.assign',

  // custom ones
  object: 'object',
  ref: 'ref',
  catchall: 'catchall',
  matches: 'faker.helpers.fromRegExp',
  email: 'faker.internet.email',
  firstName: 'faker.person.firstName',
  lastName: 'faker.person.lastName',
  password: 'faker.internet.password',
  phone: 'faker.phone.number',
} satisfies { [K in keyof FakerMetaMapper]: string }

type FakerMetaBase<T> = {
  keyword: FakerKeyword
  args: T
}

export type FakerMeta =
  | { keyword: string }
  | FakerMetaMapper[keyof FakerMetaMapper]

export function isKeyword<T extends FakerMeta, K extends keyof FakerMetaMapper>(meta: T, keyword: K): meta is Extract<T, FakerMetaMapper[K]> {
  return meta.keyword === keyword
}

/**
 * @link based on https://github.com/cellular/oazapfts/blob/7ba226ebb15374e8483cc53e7532f1663179a22c/src/codegen/generate.ts#L398
 */

function fakerKeywordSorter(a: FakerMeta, b: FakerMeta) {
  if (b.keyword === 'null') {
    return -1
  }

  return 0
}

function joinItems(items: string[]): string {
  switch (items.length) {
    case 0:
      return 'undefined'
    case 1:
      return items[0]!
    default:
      return `${fakerKeywordMapper.union}([${items.join(',')}])`
  }
}

export function parseFakerMeta(
  item: FakerMeta = {} as FakerMeta,
  { mapper = fakerKeywordMapper, withOverride }: { mapper?: Record<FakerKeyword, string>; withOverride?: boolean } = {},
): string {
  const value = mapper[item.keyword as keyof typeof mapper]

  if (isKeyword(item, fakerKeywords.tuple) || isKeyword(item, fakerKeywords.array) || isKeyword(item, fakerKeywords.union)) {
    return `${value}(${
      Array.isArray(item.args)
        ? `[${item.args.map((orItem) => parseFakerMeta(orItem, { mapper })).join(',')}]`
        : parseFakerMeta(item.args)
    }) as any`
  }

  if (isKeyword(item, fakerKeywords.and)) {
    return `${value}({},${
      Array.isArray(item.args) ? `${item.args.map((andItem) => parseFakerMeta(andItem, { mapper })).join(',')}` : parseFakerMeta(item.args)
    })`
  }

  if (isKeyword(item, fakerKeywords.enum)) {
    return `${value}(${Array.isArray(item.args) ? `${item.args.join(',')}` : parseFakerMeta(item.args)})`
  }

  if (isKeyword(item, fakerKeywords.catchall)) {
    throw new Error('catchall is not implemented')
  }

  if (isKeyword(item, fakerKeywords.object)) {
    const argsObject = Object.entries(item.args?.entries || '{}')
      .filter((item) => {
        const schema = item[1]
        return schema && typeof schema.map === 'function'
      })
      .map((item) => {
        const name = item[0]
        const schema = item[1]
        return `"${name}": ${
          joinItems(
            schema
              .sort(fakerKeywordSorter)
              .map((item) => parseFakerMeta(item, { mapper })),
          )
        }`
      })
      .join(',')

    return `{${argsObject}}`
  }

  // custom type
  if (isKeyword(item, fakerKeywords.ref)) {
    if (!item.args?.name) {
      throw new Error(`Name not defined for keyword ${item.keyword}`)
    }

    if (withOverride) {
      return `${item.args.name}(override)`
    }
    return `${item.args.name}()`
  }

  if (isKeyword(item, fakerKeywords.null) || isKeyword(item, fakerKeywords.undefined) || isKeyword(item, fakerKeywords.any)) {
    return value
  }

  if (isKeyword(item, fakerKeywords.matches)) {
    const options = (item as FakerMetaBase<unknown>).args as string
    let regex
    try {
      regex = new RegExp(options)
    } catch (_e) {
      regex = JSON.stringify(options)
    }

    return `${value}(${regex ?? ''})`
  }

  if (item.keyword in mapper) {
    const options = JSON.stringify((item as FakerMetaBase<unknown>).args)
    return `${value}(${options ?? ''})`
  }

  return '""'
}

export function fakerParser(
  items: FakerMeta[],
  options: { seed?: number | number[]; mapper?: Record<FakerKeyword, string>; name: string; typeName?: string | null },
): string {
  const fakerText = joinItems(
    items.map((item) => parseFakerMeta(item, { mapper: { ...fakerKeywordMapper, ...options.mapper }, withOverride: true })),
  )

  let fakerDefaultOverride: '' | '[]' | '{}' = ''
  let fakerTextWithOverride = fakerText

  if (fakerText.startsWith('{')) {
    fakerDefaultOverride = '{}'
    fakerTextWithOverride = `{
  ...${fakerText},
  ...override
}`
  }

  if (fakerText.startsWith(fakerKeywordMapper.array)) {
    fakerDefaultOverride = '[]'
    fakerTextWithOverride = `[
      ...${fakerText},
      ...override
    ]`
  }

  return `
export function ${options.name}(${
    fakerDefaultOverride
      ? `override: NonNullable<Partial<${options.typeName}>> = ${fakerDefaultOverride}`
      : `override?: NonNullable<Partial<${options.typeName}>>`
  })${options.typeName ? `: NonNullable<${options.typeName}>` : ''} {
  ${options.seed ? `faker.seed(${JSON.stringify(options.seed)})` : ''}
  return ${fakerTextWithOverride};
}
  `
}
