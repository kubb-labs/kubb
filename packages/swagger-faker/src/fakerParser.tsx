import transformers, { createJSDocBlockText } from '@kubb/core/transformers'
import { isKeyword, schemaKeywords } from '@kubb/swagger'

import type { Schema, SchemaKeywordBase, SchemaMapper } from '@kubb/swagger'

export const fakerKeywordMapper = {
  any: 'undefined',
  unknown: 'unknown',
  number: 'faker.number.float',
  integer: 'faker.number.int',
  string: 'faker.string.alpha',
  boolean: 'faker.datatype.boolean',
  undefined: 'undefined',
  null: 'null',
  array: 'faker.helpers.arrayElements',
  tuple: 'faker.helpers.arrayElements',
  enum: 'faker.helpers.arrayElement<any>',
  union: 'faker.helpers.arrayElement',
  date: 'faker.date.anytime',
  datetime: 'faker.date.anytime().toISOString',
  uuid: 'faker.string.uuid',
  url: 'faker.internet.url',
  and: 'Object.assign',
  object: 'object',
  ref: 'ref',
  matches: 'faker.helpers.fromRegExp',
  email: 'faker.internet.email',
  firstName: 'faker.person.firstName',
  lastName: 'faker.person.lastName',
  password: 'faker.internet.password',
  phone: 'faker.phone.number',
  blob: undefined,
  default: undefined,
  describe: undefined,
  lazy: undefined,
  const: undefined,
  max: undefined,
  min: undefined,
  nullable: undefined,
  nullish: undefined,
  optional: undefined,
  readOnly: undefined,
  strict: undefined,
  deprecated: undefined,
  example: undefined,
  type: undefined,
  format: undefined,
  catchall: undefined,
} satisfies SchemaMapper

/**
 * @link based on https://github.com/cellular/oazapfts/blob/7ba226ebb15374e8483cc53e7532f1663179a22c/src/codegen/generate.ts#L398
 */

function schemaKeywordsorter(a: Schema, b: Schema) {
  if (b.keyword === 'null') {
    return -1
  }

  return 0
}

function joinItems(items: string[]): string {
  switch (items.length) {
    case 0:
      return 'undefined'
    case 1:
      return items[0]!
    default:
      return `${fakerKeywordMapper.union}([${items.join(',')}])`
  }
}

type ParserOptions = {
  name: string
  typeName?: string
  description?: string

  seed?: number | number[]
  withOverride?: boolean
  mapper?: typeof fakerKeywordMapper
}

export function parseFakerMeta(item: Schema, options: ParserOptions): string | undefined {
  const mapper = options.mapper || fakerKeywordMapper
  const value = mapper[item.keyword as keyof typeof mapper]

  if (!value) {
    return undefined
  }

  if (isKeyword(item, schemaKeywords.union)) {
    return `${value}([${item.args
      .map((orItem) => parseFakerMeta(orItem, { ...options, withOverride: false }))
      .filter(Boolean)
      .join(',')}]) as any`
  }

  if (isKeyword(item, schemaKeywords.and)) {
    return `${value}({},${item.args
      .map((andItem) => parseFakerMeta(andItem, { ...options, withOverride: false }))
      .filter(Boolean)
      .join(',')})`
  }

  if (isKeyword(item, schemaKeywords.array)) {
    return `${value}([${item.args.items
      .map((orItem) => parseFakerMeta(orItem, { ...options, withOverride: false }))
      .filter(Boolean)
      .join(',')}]) as any`
  }

  if (isKeyword(item, schemaKeywords.enum)) {
    return `${value}([${item.args.items
      .map((item) => {
        if (item.format === 'number') {
          return item.name
        }
        return transformers.stringify(item.name)
      })
      .join(', ')}])`
  }

  if (isKeyword(item, schemaKeywords.ref)) {
    if (!item.args?.name) {
      throw new Error(`Name not defined for keyword ${item.keyword}`)
    }

    if (options.withOverride) {
      return `${item.args.name}(override)`
    }

    return `${item.args.name}()`
  }

  if (isKeyword(item, schemaKeywords.object)) {
    const argsObject = Object.entries(item.args?.properties || {})
      .filter((item) => {
        const schema = item[1]
        return schema && typeof schema.map === 'function'
      })
      .map((item) => {
        const name = item[0]
        const schema = item[1]

        return `"${name}": ${joinItems(
          schema
            .sort(schemaKeywordsorter)
            .map((item) => parseFakerMeta(item, { ...options, withOverride: false }))
            .filter(Boolean),
        )}`
      })
      .join(',')

    return `{${argsObject}}`
  }

  if (isKeyword(item, schemaKeywords.tuple)) {
    return `${value}(${
      Array.isArray(item.args)
        ? `[${item.args
            .map((orItem) => parseFakerMeta(orItem, { ...options, withOverride: false }))
            .filter(Boolean)
            .join(',')}]`
        : parseFakerMeta(item.args, { ...options, withOverride: false })
    }) as any`
  }

  if (isKeyword(item, schemaKeywords.const)) {
    if (item.args.format === 'number') {
      return item.args.name?.toString()
    }
    return transformers.stringify(item.args.value)
  }

  if (isKeyword(item, schemaKeywords.matches)) {
    if (item.args) {
      return `${value}(${transformers.toRegExpString(item.args)})`
    }
  }

  if (isKeyword(item, schemaKeywords.null) || isKeyword(item, schemaKeywords.undefined) || isKeyword(item, schemaKeywords.any)) {
    return value || ''
  }

  if (item.keyword in mapper) {
    const options = JSON.stringify((item as SchemaKeywordBase<unknown>).args)
    return `${value}(${options ?? ''})`
  }

  return undefined
}

export function fakerParser(schemas: Schema[], options: ParserOptions): string {
  const fakerText = joinItems(schemas.map((item) => parseFakerMeta(item, { ...options, withOverride: true })).filter(Boolean))

  let fakerDefaultOverride: '' | '[]' | '{}' = ''
  let fakerTextWithOverride = fakerText

  if (fakerText.startsWith('{')) {
    fakerDefaultOverride = '{}'
    fakerTextWithOverride = `{
  ...${fakerText},
  ...override
}`
  }

  if (fakerText.startsWith(fakerKeywordMapper.array)) {
    fakerDefaultOverride = '[]'
    fakerTextWithOverride = `[
      ...${fakerText},
      ...override
    ]`
  }

  const JSDoc = createJSDocBlockText({
    comments: [options.description ? `@description ${transformers.jsStringEscape(options.description)}` : undefined].filter(Boolean),
  })

  return `
${JSDoc}
export function ${options.name}(${
    fakerDefaultOverride
      ? `override: NonNullable<Partial<${options.typeName}>> = ${fakerDefaultOverride}`
      : `override?: NonNullable<Partial<${options.typeName}>>`
  })${options.typeName ? `: NonNullable<${options.typeName}>` : ''} {
  ${options.seed ? `faker.seed(${JSON.stringify(options.seed)})` : ''}
  return ${fakerTextWithOverride}
}\n`

  // const root = createRoot()

  // root.render(
  //   <Function
  //     export
  //     name={options.name}
  //     JSDoc={{ comments: [describeSchema ? `@description ${transformers.stringify(describeSchema.args)}` : undefined].filter(Boolean) }}
  //     params={fakerDefaultOverride
  //       ? `override: NonNullable<Partial<${options.typeName}>> = ${fakerDefaultOverride}`
  //       : `override?: NonNullable<Partial<${options.typeName}>>`}
  //     returnType={options.typeName ? `NonNullable<${options.typeName}>` : ''}
  //   >
  //     {options.seed ? `faker.seed(${JSON.stringify(options.seed)})` : ''}

  //     {`return ${fakerTextWithOverride}`}
  //   </Function>,
  // )

  // return root.output
}
