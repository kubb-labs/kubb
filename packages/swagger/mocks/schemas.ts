import { schemaKeywords, type Schema } from '../src/SchemaMapper'

const basic: Array<{ name: string; schema: Schema }> = [
  {
    name: 'any',
    schema: {
      keyword: schemaKeywords.any,
    },
  },
  {
    name: 'unknown',
    schema: {
      keyword: schemaKeywords.unknown,
    },
  },
  {
    name: schemaKeywords.string,
    schema: {
      keyword: schemaKeywords.string,
    },
  },
  {
    name: schemaKeywords.number,
    schema: {
      keyword: schemaKeywords.number,
    },
  },
  {
    name: 'boolean',
    schema: {
      keyword: schemaKeywords.boolean,
    },
  },
  {
    name: 'date',
    schema: {
      keyword: schemaKeywords.date,
    },
  },
  {
    name: 'datetime',
    schema: {
      keyword: schemaKeywords.datetime,
    },
  },
  {
    name: 'nullable',
    schema: {
      keyword: schemaKeywords.nullable,
    },
  },
  {
    name: 'undefined',
    schema: {
      keyword: schemaKeywords.undefined,
    },
  },
  {
    name: schemaKeywords.min,
    schema: {
      keyword: schemaKeywords.min,
      args: 2,
    },
  },
  {
    name: 'max',
    schema: {
      keyword: schemaKeywords.max,
      args: 2,
    },
  },
  {
    name: 'matchesReg',
    schema: {
      keyword: schemaKeywords.matches,
      args: '/node_modules/', // pure regexp
    },
  },
  {
    name: 'matches',
    schema: {
      keyword: schemaKeywords.matches,
      args: '^[A-Z]{2}$',
    },
  },
  {
    name: 'ref',
    schema: {
      keyword: schemaKeywords.ref,
      args: {
        name: 'Pet',
        path: './pet.ts',
      },
    },
  },
  {
    name: 'enum',
    schema: {
      keyword: schemaKeywords.enum,
      args: {
        name: 'enum',
        typeName: 'Enum',
        asConst: false,
        items: [
          { name: 'A', value: 'A', format: schemaKeywords.string },
          { name: 'B', value: 'B', format: schemaKeywords.string },
          { name: 'C', value: 'C', format: schemaKeywords.string },
          { name: 2, value: 2, format: schemaKeywords.number },
        ],
      },
    },
  },
  {
    name: 'tuple',
    schema: {
      keyword: schemaKeywords.tuple,
      args: [],
    },
  },
  {
    name: 'tupleMulti',
    schema: {
      keyword: schemaKeywords.tuple,
      args: [{ keyword: schemaKeywords.string }, { keyword: schemaKeywords.number }],
    },
  },
  {
    name: 'array',
    schema: {
      keyword: schemaKeywords.array,
      args: {
        items: [
          {
            keyword: schemaKeywords.union,
            args: [{ keyword: schemaKeywords.number }, { keyword: schemaKeywords.string }],
          },
        ],
      },
    },
  },
  {
    name: 'arrayEmpty',
    schema: {
      keyword: schemaKeywords.array,
      args: {
        items: [],
      },
    },
  },
  {
    name: 'arrayRef',
    schema: {
      keyword: schemaKeywords.array,
      args: {
        items: [
          {
            keyword: schemaKeywords.ref,
            args: { name: 'Pet', path: './pet.ts' },
          },
        ],
      },
    },
  },
  {
    name: 'arrayAdvanced',
    schema: {
      keyword: schemaKeywords.array,
      args: {
        items: [{ keyword: schemaKeywords.min, args: 1 }, { keyword: schemaKeywords.number }],
        min: 3,
        max: 10,
      },
    },
  },
  {
    name: 'union',
    schema: {
      keyword: schemaKeywords.union,
      args: [{ keyword: schemaKeywords.string }, { keyword: schemaKeywords.number }],
    },
  },
  {
    name: 'unionOne',
    schema: {
      keyword: schemaKeywords.union,
      args: [{ keyword: schemaKeywords.string }],
    },
  },
  {
    name: 'catchall',
    schema: {
      keyword: schemaKeywords.object,
      args: {
        properties: {},
        additionalProperties: [
          {
            keyword: schemaKeywords.ref,
            args: { name: 'Pet', path: './Pet.ts' },
          },
        ],
      },
    },
  },
  {
    name: 'and',
    schema: {
      keyword: schemaKeywords.and,
      args: [{ keyword: schemaKeywords.string }, { keyword: schemaKeywords.number }],
    },
  },
  {
    name: 'object',
    schema: {
      keyword: schemaKeywords.object,
      args: {
        properties: {
          firstName: [{ keyword: schemaKeywords.string }, { keyword: schemaKeywords.min, args: 2 }],
          address: [{ keyword: schemaKeywords.string }, { keyword: schemaKeywords.nullable }, { keyword: schemaKeywords.describe, args: '"Your address"' }],
        },
        additionalProperties: [],
      },
    },
  },
  {
    name: 'objectOptional',
    schema: {
      keyword: schemaKeywords.object,
      args: {
        properties: {
          firstName: [{ keyword: schemaKeywords.string }, { keyword: schemaKeywords.optional }, { keyword: schemaKeywords.min, args: 2 }],
          address: [{ keyword: schemaKeywords.string }, { keyword: schemaKeywords.nullable }, { keyword: schemaKeywords.describe, args: '"Your address"' }],
        },
        additionalProperties: [],
      },
    },
  },
  {
    name: 'objectAnd',
    schema: {
      keyword: schemaKeywords.object,
      args: {
        properties: {
          firstName: [
            { keyword: schemaKeywords.deprecated },
            { keyword: schemaKeywords.default, args: 'test' },
            {
              keyword: schemaKeywords.min,
              args: 2,
            },
            {
              keyword: schemaKeywords.string,
            },
          ],
          age: [
            { keyword: schemaKeywords.example, args: '2' },
            { keyword: schemaKeywords.default, args: 2 },
            {
              keyword: schemaKeywords.min,
              args: 2,
            },
            {
              keyword: schemaKeywords.number,
            },
          ],
          address: [
            {
              keyword: schemaKeywords.and,
              args: [{ keyword: schemaKeywords.string }, { keyword: schemaKeywords.number }],
            },
            { keyword: schemaKeywords.nullable },
            { keyword: schemaKeywords.describe, args: 'Your address' },
          ],
        },
        additionalProperties: [],
      },
    },
  },
  {
    name: 'objectEmpty',
    schema: {
      keyword: schemaKeywords.object,
      args: {
        properties: {},
        additionalProperties: [],
      },
    },
  },
  {
    name: 'default',
    schema: {
      keyword: schemaKeywords.default,
    },
  },
  {
    name: 'default',
    schema: {
      keyword: schemaKeywords.default,
      args: 'default',
    },
  },
  {
    name: 'blob',
    schema: {
      keyword: schemaKeywords.blob,
    },
  },
]

const full: Array<{ name: string; schema: Schema[] }> = [
  {
    name: 'Upload',
    schema: [
      {
        keyword: schemaKeywords.blob,
      },
    ],
  },
  {
    name: 'Object',
    schema: [
      { keyword: schemaKeywords.nullable },
      { keyword: schemaKeywords.describe, args: 'Your address' },
      {
        keyword: schemaKeywords.object,
        args: {
          properties: {
            firstName: [
              { keyword: schemaKeywords.deprecated },
              { keyword: schemaKeywords.default, args: 'test' },
              {
                keyword: schemaKeywords.min,
                args: 2,
              },
              {
                keyword: schemaKeywords.string,
              },
            ],
            age: [
              { keyword: schemaKeywords.example, args: '2' },
              { keyword: schemaKeywords.default, args: 2 },
              {
                keyword: schemaKeywords.min,
                args: 2,
              },
              {
                keyword: schemaKeywords.number,
              },
            ],
            address: [
              {
                keyword: schemaKeywords.and,
                args: [{ keyword: schemaKeywords.string }, { keyword: schemaKeywords.number }],
              },
              { keyword: schemaKeywords.nullable },
              { keyword: schemaKeywords.describe, args: 'Your address' },
            ],
          },
          additionalProperties: [],
        },
      },
    ],
  },
  {
    name: 'Order',
    schema: [
      {
        keyword: schemaKeywords.type,
        args: 'object',
      },
      {
        keyword: schemaKeywords.object,
        args: {
          properties: {
            status: [
              {
                keyword: schemaKeywords.enum,
                args: {
                  name: 'orderStatus',
                  asConst: false,
                  typeName: 'OrderStatus',
                  items: [
                    { name: 'Placed', value: 'placed', format: 'string' },
                    { name: 'Approved', value: 'approved', format: 'string' },
                  ],
                },
              },
            ],
          },
          additionalProperties: [],
        },
      },
    ],
  },
  {
    name: 'Record',
    schema: [
      {
        keyword: schemaKeywords.object,
        args: {
          properties: {},
          additionalProperties: [
            {
              keyword: schemaKeywords.integer,
            },
            {
              keyword: schemaKeywords.type,
              args: 'integer',
            },
            {
              keyword: schemaKeywords.format,
              args: 'int32',
            },
            {
              keyword: schemaKeywords.optional,
            },
          ],
        },
      },
      {
        keyword: 'type',
        args: 'object',
      },
      {
        keyword: 'optional',
      },
    ],
  },
]

export const schemas = {
  basic,
  full,
}
