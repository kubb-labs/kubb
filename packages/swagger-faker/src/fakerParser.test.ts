import { schemas } from '@kubb/swagger/mocks/schemas.js'

import { fakerParser, parseFakerMeta } from './fakerParser.tsx'

describe('parseFakerMeta', () => {
  test.each(schemas.basic)('$name', ({ name, schema }) => {
    const text = parseFakerMeta(schema, { name })
    expect(text).toMatchSnapshot()
  })
})

describe('fakerParser', () => {
  test.each(schemas.full)('$name', ({ schema, name }) => {
    const text = fakerParser(schema, { name })
    expect(text).toMatchSnapshot()
  })
})
