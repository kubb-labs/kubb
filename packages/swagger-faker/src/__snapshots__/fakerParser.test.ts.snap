// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fakerParser > 'Object' 1`] = `
"

export function Object() {
  
  return {"firstName": faker.string.alpha(),"age": faker.number.float(),"address": Object.assign({}, faker.string.alpha(), faker.number.float())}
}
"
`;

exports[`fakerParser > 'Order' 1`] = `
"

export function Order() {
  
  return {"status": faker.helpers.arrayElement<any>(["Placed", "Approved"])}
}
"
`;

exports[`fakerParser > 'Record' 1`] = `
"

export function Record() {
  
  return {}
}
"
`;

exports[`fakerParser > 'Upload' 1`] = `
"

export function Upload() {
  
  return undefined
}
"
`;

exports[`parseFakerMeta > 'and' 1`] = `"Object.assign({}, faker.string.alpha(), faker.number.float())"`;

exports[`parseFakerMeta > 'any' 1`] = `"undefined"`;

exports[`parseFakerMeta > 'array' 1`] = `"faker.helpers.arrayElements([faker.helpers.arrayElement<any>([faker.number.float(), faker.string.alpha()])]) as any"`;

exports[`parseFakerMeta > 'arrayAdvanced' 1`] = `"faker.helpers.arrayElements([faker.number.float({ min: 1, max: 10 })]) as any"`;

exports[`parseFakerMeta > 'arrayEmpty' 1`] = `"faker.helpers.arrayElements([]) as any"`;

exports[`parseFakerMeta > 'arrayRef' 1`] = `"faker.helpers.arrayElements([Pet()]) as any"`;

exports[`parseFakerMeta > 'arrayRegex' 1`] = `"faker.helpers.arrayElements([faker.helpers.fromRegExp(new RegExp('^[a-zA-Z0-9]{1,13}$')), faker.string.alpha()]) as any"`;

exports[`parseFakerMeta > 'blob' 1`] = `undefined`;

exports[`parseFakerMeta > 'boolean' 1`] = `"faker.datatype.boolean()"`;

exports[`parseFakerMeta > 'catchall' 1`] = `"{}"`;

exports[`parseFakerMeta > 'date' 1`] = `"faker.date.anytime()"`;

exports[`parseFakerMeta > 'datetime' 1`] = `"faker.date.anytime().toISOString()"`;

exports[`parseFakerMeta > 'default' 1`] = `undefined`;

exports[`parseFakerMeta > 'default' 2`] = `undefined`;

exports[`parseFakerMeta > 'enum' 1`] = `"faker.helpers.arrayElement<any>(["A", "B", "C", 2])"`;

exports[`parseFakerMeta > 'matches' 1`] = `"faker.helpers.fromRegExp(new RegExp('^[A-Z]{2}$'))"`;

exports[`parseFakerMeta > 'matchesReg' 1`] = `"faker.helpers.fromRegExp(new RegExp('node_modules'))"`;

exports[`parseFakerMeta > 'max' 1`] = `undefined`;

exports[`parseFakerMeta > 'min' 1`] = `undefined`;

exports[`parseFakerMeta > 'nullable' 1`] = `undefined`;

exports[`parseFakerMeta > 'number' 1`] = `"faker.number.float()"`;

exports[`parseFakerMeta > 'object' 1`] = `"{"firstName": faker.string.alpha(),"address": faker.string.alpha()}"`;

exports[`parseFakerMeta > 'objectAnd' 1`] = `"{"firstName": faker.string.alpha(),"age": faker.number.float(),"address": Object.assign({}, faker.string.alpha(), faker.number.float())}"`;

exports[`parseFakerMeta > 'objectArray' 1`] = `"{"ids": faker.helpers.arrayElements([faker.helpers.fromRegExp(new RegExp('^[a-zA-Z0-9]{1,13}$')), faker.string.alpha()]) as any}"`;

exports[`parseFakerMeta > 'objectEmpty' 1`] = `"{}"`;

exports[`parseFakerMeta > 'objectEnum' 1`] = `"{"version": faker.helpers.arrayElement<any>(["A", "B", "C", 2])}"`;

exports[`parseFakerMeta > 'objectOptional' 1`] = `"{"firstName": faker.string.alpha(),"address": faker.string.alpha()}"`;

exports[`parseFakerMeta > 'ref' 1`] = `"Pet()"`;

exports[`parseFakerMeta > 'string' 1`] = `"faker.string.alpha()"`;

exports[`parseFakerMeta > 'stringOffset' 1`] = `"faker.date.anytime().toISOString()"`;

exports[`parseFakerMeta > 'tuple' 1`] = `"faker.helpers.arrayElements([]) as any"`;

exports[`parseFakerMeta > 'tupleMulti' 1`] = `"faker.helpers.arrayElements([faker.string.alpha(), faker.number.float()]) as any"`;

exports[`parseFakerMeta > 'undefined' 1`] = `"undefined"`;

exports[`parseFakerMeta > 'union' 1`] = `"faker.helpers.arrayElement<any>([faker.string.alpha(), faker.number.float()])"`;

exports[`parseFakerMeta > 'unionOne' 1`] = `"faker.helpers.arrayElement<any>([faker.string.alpha()])"`;

exports[`parseFakerMeta > 'unknown' 1`] = `"unknown"`;
